cmd/bootstrap/cmd/seal.go

```diff
@@ -78,3 +81,22 @@ func constructRootResultAndSeal(
 
 	return result, seal
 }
+
+// rootEpochTargetEndTime computes the target end time for the given epoch, using the given config.
+// CAUTION: the variables `flagEpochTimingRefCounter`, `flagEpochTimingDuration`, and
+// `flagEpochTimingRefTimestamp` must contain proper values. You can either specify a value for
+// each config parameter or use the function `validateOrPopulateEpochTimingConfig()` to populate the variables
+// from defaults.
+func rootEpochTargetEndTime() uint64 {
+	if flagEpochTimingRefTimestamp == 0 || flagEpochTimingDuration == 0 {
+		panic("invalid epoch timing config: must specify ALL of --epoch-target-end-time-ref-counter, --epoch-target-end-time-ref-timestamp, and --epoch-target-end-time-duration")
+	}
+	if flagEpochCounter < flagEpochTimingRefCounter {
+		panic("invalid epoch timing config: reference epoch counter must be before root epoch counter")
+	}
+	targetEndTime := flagEpochTimingRefTimestamp + (flagEpochCounter-flagEpochTimingRefCounter)*flagEpochTimingDuration
+	if targetEndTime < uint64(time.Now().Unix()) {
```

Comment:
This is largely academic 😅 
```suggestion
	if targetEndTime <= uint64(time.Now().Unix()) {
```